File: sdks/php/examples/authentication.php
<?php

declare(strict_types=1);

/**
 * Authentication Examples for Synap PHP SDK
 * 
 * This example demonstrates how to use authentication with the Synap PHP SDK,
 * including Basic Auth and API Key authentication.
 */

require_once __DIR__ . '/../vendor/autoload.php';

use Synap\SDK\SynapClient;
use Synap\SDK\SynapConfig;

function exampleBasicAuth(): void
{
    echo "=== Basic Auth Example ===\n\n";
    
    // Create config with Basic Auth credentials
    $config = SynapConfig::create('http://localhost:15500')
        ->withBasicAuth('root', 'root');
    
    $client = new SynapClient($config);
    
    echo "Testing connection with Basic Auth...\n";
    
    // Perform operations - authentication is automatic
    $client->kv()->set('test:key', 'test_value');
    $value = $client->kv()->get('test:key');
    echo "✅ Successfully set and retrieved value: {$value}\n";
    
    // Clean up
    $client->kv()->delete('test:key');
    echo "✅ Cleaned up test key\n\n";
}

function exampleApiKeyAuth(): void
{
    echo "=== API Key Authentication Example ===\n\n";
    
    // Create config with API Key
    $config = SynapConfig::create('http://localhost:15500')
        ->withAuthToken('your-api-key-here'); // Replace with actual API key
    
    $client = new SynapClient($config);
    
    echo "Using API key authentication...\n";
    
    // Perform operations - API key authentication is automatic
    $client->kv()->set('test:api_key', 'test_value');
    $value = $client->kv()->get('test:api_key');
    echo "✅ Successfully set and retrieved value with API key: {$value}\n";
    
    // Clean up
    $client->kv()->delete('test:api_key');
    echo "✅ Cleaned up test key\n\n";
}

function exampleBuilderPattern(): void
{
    echo "=== Builder Pattern Example ===\n\n";
    
    // Create base config
    $config = SynapConfig::create('http://localhost:15500')
        ->withTimeout(30)
        ->withBasicAuth('root', 'root');
    
    $client = new SynapClient($config);
    
    $client->kv()->set('test:builder', 'test_value');
    $value = $client->kv()->get('test:builder');
    echo "✅ Successfully used builder pattern: {$value}\n";
    
    // Clean up
    $client->kv()->delete('test:builder');
    echo "✅ Cleaned up test key\n\n";
}

function exampleSwitchAuthMethods(): void
{
    echo "=== Switching Auth Methods Example ===\n\n";
    
    // Start with Basic Auth
    $basicConfig = SynapConfig::create('http://localhost:15500')
        ->withBasicAuth('root', 'root');
    
    $basicClient = new SynapClient($basicConfig);
    $basicClient->kv()->set('test:switch', 'basic_auth');
    echo "✅ Set value using Basic Auth\n";
    
    // Switch to API Key (if you have one)
    // $apiKeyConfig = SynapConfig::create('http://localhost:15500')
    //     ->withAuthToken('your-api-key');
    // $apiKeyClient = new SynapClient($apiKeyConfig);
    // $value = $apiKeyClient->kv()->get('test:switch');
    // echo "✅ Retrieved value using API Key: {$value}\n";
    
    // Clean up
    $basicClient->kv()->delete('test:switch');
    echo "✅ Cleaned up test key\n\n";
}

// Main execution
echo "🔐 Synap PHP SDK - Authentication Examples\n";
echo str_repeat('=', 50) . "\n\n";

try {
    exampleBasicAuth();
    exampleApiKeyAuth();
    exampleBuilderPattern();
    exampleSwitchAuthMethods();
    
    echo str_repeat('=', 50) . "\n";
    echo "✅ All authentication examples completed successfully!\n";
} catch (Exception $e) {
    echo "❌ Error: {$e->getMessage()}\n";
    exit(1);
}

File: sdks/rust/examples/authentication.rs
//! Authentication Examples for Synap Rust SDK
//!
//! This example demonstrates how to use authentication with the Synap Rust SDK,
//! including Basic Auth and API Key authentication.

use std::time::Duration;
use synap_sdk::client::{SynapClient, SynapConfig};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("🔐 Synap Rust SDK - Authentication Examples\n");
    println!("{}", "=".repeat(50));
    println!();

    // Example 1: Basic Auth
    example_basic_auth().await?;

    // Example 2: API Key Auth
    example_api_key_auth().await?;

    // Example 3: Builder Pattern
    example_builder_pattern().await?;

    // Example 4: Switch Auth Methods
    example_switch_auth_methods().await?;

    println!("{}", "=".repeat(50));
    println!("✅ All authentication examples completed successfully!");

    Ok(())
}

async fn example_basic_auth() -> Result<(), Box<dyn std::error::Error>> {
    println!("=== Basic Auth Example ===\n");

    // Create config with Basic Auth credentials
    let config = SynapConfig::new("http://localhost:15500").with_basic_auth("root", "root");

    let client = SynapClient::new(config)?;

    println!("Testing connection with Basic Auth...");

    // Perform operations - authentication is automatic
    client.kv().set("test:key", "test_value", None).await?;
    let value: Option<Vec<u8>> = client.kv().get("test:key").await?;

    if let Some(bytes) = value {
        let value_str = String::from_utf8_lossy(&bytes);
        println!("✅ Successfully set and retrieved value: {}", value_str);
    }

    // Clean up
    client.kv().delete("test:key").await?;
    println!("✅ Cleaned up test key\n");

    Ok(())
}

async fn example_api_key_auth() -> Result<(), Box<dyn std::error::Error>> {
    println!("=== API Key Authentication Example ===\n");

    // Create config with API Key
    let config = SynapConfig::new("http://localhost:15500").with_auth_token("your-api-key-here"); // Replace with actual API key

    let client = SynapClient::new(config)?;

    println!("Using API key authentication...");

    // Perform operations - API key authentication is automatic
    client.kv().set("test:api_key", "test_value", None).await?;
    let value: Option<Vec<u8>> = client.kv().get("test:api_key").await?;

    if let Some(bytes) = value {
        let value_str = String::from_utf8_lossy(&bytes);
        println!(
            "✅ Successfully set and retrieved value with API key: {}",
            value_str
        );
    }

    // Clean up
    client.kv().delete("test:api_key").await?;
    println!("✅ Cleaned up test key\n");

    Ok(())
}

async fn example_builder_pattern() -> Result<(), Box<dyn std::error::Error>> {
    println!("=== Builder Pattern Example ===\n");

    // Create base config
    let config = SynapConfig::new("http://localhost:15500")
        .with_timeout(Duration::from_secs(30))
        .with_basic_auth("root", "root");

    let client = SynapClient::new(config)?;

    client.kv().set("test:builder", "test_value", None).await?;
    let value: Option<Vec<u8>> = client.kv().get("test:builder").await?;

    if let Some(bytes) = value {
        let value_str = String::from_utf8_lossy(&bytes);
        println!("✅ Successfully used builder pattern: {}", value_str);
    }

    // Clean up
    client.kv().delete("test:builder").await?;
    println!("✅ Cleaned up test key\n");

    Ok(())
}

async fn example_switch_auth_methods() -> Result<(), Box<dyn std::error::Error>> {
    println!("=== Switching Auth Methods Example ===\n");

    // Start with Basic Auth
    let basic_config = SynapConfig::new("http://localhost:15500").with_basic_auth("root", "root");

    let basic_client = SynapClient::new(basic_config)?;
    basic_client
        .kv()
        .set("test:switch", "basic_auth", None)
        .await?;
    println!("✅ Set value using Basic Auth");

    // Switch to API Key (if you have one)
    // let api_key_config = SynapConfig::new("http://localhost:15500")
    //     .with_auth_token("your-api-key");
    // let api_key_client = SynapClient::new(api_key_config)?;
    // let value = api_key_client.kv().get("test:switch").await?;
    // if let Some(bytes) = value {
    //     let value_str = String::from_utf8_lossy(&bytes);
    //     println!("✅ Retrieved value using API Key: {}", value_str);
    // }

    // Clean up
    basic_client.kv().delete("test:switch").await?;
    println!("✅ Cleaned up test key\n");

    Ok(())
}
